src/dbnode/storage/index/block.go

```diff
@@ -884,6 +919,190 @@ func (b *block) addQueryResults(
 	return batch, size, err
 }
 
+// Aggregate acquires a read lock on the block so that the segments
+// are guaranteed to not be freed/released while accumulating results.
+// This allows references to the mmap'd segment data to be accumulated
+// and then copied into the results before this method returns (it is not
+// safe to return docs directly from the segments from this method, the
+// results datastructure is used to copy it every time documents are added
+// to the results datastructure). This is similar to how Query() operates.
+// NB: Aggregate is required in addition to Query, to optimise for the case
+// when we can skip going to raw documents, and instead rely on pre-aggregated
+// results via the FST underlying the index.
+func (b *block) Aggregate(
+	cancellable *resource.CancellableLifetime,
+	opts QueryOptions,
+	results AggregateResults,
+) (bool, error) {
+	b.RLock()
+	defer b.RUnlock()
+
+	if b.state == blockStateClosed {
+		return false, ErrUnableToQueryBlockClosed
+	}
+
+	segs, err := b.segmentsWithRLock()
+	if err != nil {
+		return false, err
+	}
+
+	size := results.Size()
+	batch := b.aggEntryPool.Get()
+	batchSize := cap(batch)
+	if batchSize == 0 {
+		batchSize = defaultAggregateResultsEntryBatchSize
+	}
+
+	aggOpts := results.AggregateResultsOptions()
+	iterateTerms := aggOpts.Type == AggregateTagNamesAndValues
+	defer func() {
+		b.aggEntryPool.Put(batch)
+	}()
+
+	iter, err := newFieldsAndTermsIterator(nil, fieldsAndTermsIteratorOpts{})
+	if err != nil {
+		return false, err
+	}
+
+	for _, s := range segs {
+		if opts.LimitExceeded(size) {
+			break
+		}
+
+		err = iter.Reset(s, fieldsAndTermsIteratorOpts{
```

Comment:
nit: worth pulling the opts out of the loop?

Reply:
allocation wise it's a stack alloc either way so no bigge. i prefer how reads, a little more declarative.